theano/tensor/opt.py

```diff
@@ -3806,6 +3806,29 @@ def local_div_switch_sink(node):
     return False
 
 
+# Merge add/sub/mul/div/minimum/maximum of two switches sharing the same
+# condition, to enable further simplification of their branches
+# Example: switch(c, a, b) + switch(c, x, y) -> switch(c, a+x, b+y)
+# TODO: generalize to all elemwise Ops? generalize to Ops with 3+ inputs?
+for _op_name in ('add', 'sub', 'mul', 'true_div', 'int_div', 'floor_div',
```

Comment:
Code written to try to replace this without PatternSub:

```
@register_canonicalize
@gof.local_optimizer([T.Elemwise])
def local_merge_switch(node):
    if not isinstance(node.op T.Elemwise) or not isinstance(node.op.scalar_op, (scal.Add, scal.Sub, scal.Mul, scal.TrueDiv, scal.IntDiv, scal.FloorDir, ....):
        return
# all inputs must be switch
    if not all([i.owner and isinstance(i.owner.op, T.Elemwise) and isinstance(i.owner.op.scalar_op, scal.Switch) for i in node.inputs]):
        return
# all switch cond must be the same
    cond = node.inputs[0].owner.inputs[0]
    if not all([i is cond for i in node.inputs[1:]]):
        return
    return T.switch(cond,
                            node.op(*[s.owner.inputs[1] for s in node.inputs]),
                            node.op(*[s.owner.inputs[2] for s in node.inputs]))]
```

I think this version work for more then 2 inputs.